docs/Custom_Operation_for_GPUs.md

```diff
@@ -0,0 +1,2323 @@
+
+
+# rms_norm_bwd was previously defined as
+#
+# def rms_norm_bwd(g, invvar, x, weight, eps):
+#     grad_input, grad_weight, part_grad = _rms_norm_bwd_p.bind(
+#         g, invvar, x, weight, eps=eps
+#     )
+#     return grad_input, grad_weight
+#
+def rms_norm_bwd(eps, res, g):
+    invvar, x, weight = res
+    grad_input, grad_weight, part_grad = _rms_norm_bwd_p.bind(
+        g, invvar, x, weight, eps=eps
+    )
+    return grad_input, grad_weight
+```
+
+`rms_norm_fwd` now returns an extra output `(invvar, x, weight)` for the residual data and `rms_norm_bwd` takes `eps`, `res`, and `g` as the parameters.
+
+Once the relationship between `rms_norm_fwd` and `rms_norm_bwd` is estabilished through `jax.custom_vjp`, JAX will ensure that the residual data from `rms_norm_fwd` is passed to `rms_norm_bwd` as `res` for backward operation.
+For non-differentiable parameters such as `eps`, JAX ensures that they are passed to the backward operation before the residual data.  That's why `eps` precedes `res` in the parameter list of `rms_norm_bwd`.
+
+Now that `rms_norm_fwd` returns the residual data, which is not needed for simple RMS normalization operation, we define a wrapper around it, `rms_norm`.  It simply calls `rms_norm_fwd` and returns only `output`.  Note that `rms_norm` is annotated with `@partial(jax.custom_vjp, nondiff_argnums=(2,))` and we are passing `rms_norm_fwd` and `rms_norm_bwd` to `rms_norm.defvjp`.  It teaches JAX that, when `rms_norm` is differentiated, `rms_norm_fwd` is to be used for forward operation, and `rms_norm_bwd` is to be used for backward operation.
+
+See [Custom derivative rules for JAX-transformable Python functions](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html#use-jax-custom-vjp-to-define-custom-reverse-mode-only-rules) for more information on `jax.custom_vjp`.
+
+```python
+@partial(jax.custom_vjp, nondiff_argnums=(2,))
+def rms_norm(x, weight, eps=1e-05):
+    output, _ = rms_norm_fwd(x, weight, eps=eps)
+    return output
+
+
+rms_norm.defvjp(rms_norm_fwd, rms_norm_bwd)
+```
+
+With the refinement we have made, the backward operation test works with a modification: `loss` now calls `rms_norm` instead of `rms_norm_fwd`.
+
+```python
+def loss(x, weight):
+    predictions = rms_norm(x, weight)
+    return -jax.numpy.mean(predictions**2)
+
+
+loss_grad = jax.grad(loss)
+out = loss_grad(x, weight)
+```
+
+# Let's test it on multiple devices
+
+We are using `jax.experimental.pjit.pjit` for parallel execution on multiple devices, and we produce reference values with sequential execution on a single device.
+
+## Test forward function
+
+Let's first test the forward operation on multiple devices.  We are creating a simple 1D mesh and sharding `x` on all devices.
+
+```python
+import numpy
+from jax.experimental.maps import Mesh
+from jax.experimental.pjit import PartitionSpec, pjit
+
+
+mesh = Mesh(numpy.array(jax.local_devices()).reshape(-1), ("a",))
```

Comment:
s/"a"/"x" since x is more conventional

Reply:
Mesh(jax.local_devices(), 'x') should work?

Reply:
maybe `with Mesh(jax.local_devices(), 'x'):`?

Reply:
Fixed